spec/parser_spec.lua

```diff
@@ -579,6 +579,31 @@ describe("Pallene parser", function()
         })
     end)
 
+    it("can parse for-in loops", function()
+        assert_statements_ast([[
+            for k, v in ipairs(t) do
+                x = v
+            end
+        ]], {
+            { _tag = "ast.Stat.ForIn",
+              block = {
+                stats = {
+                  { _tag = "ast.Stat.Assign",
+                    exps = { { var = { _tag = "ast.Var.Name", name = "v" } } },
+                    vars = { { _tag = "ast.Var.Name", name = "x" } } } } },
+              decls = {
+                { _tag = "ast.Decl.Decl", name = "k", type = false },
+                { _tag = "ast.Decl.Decl", name = "v", type = false }, },
+              exps =  {
+                { _tag = "ast.Exp.CallFunc", exp  = { _tag = "ast.Exp.Var",
+                  var = { _tag = "ast.Var.Name", name = "ipairs" } },
+                  args = {
+                    { _tag = "ast.Exp.Var" ,
+                    var = { name = "t" } } } } }
+            }
+        })
+    end)
```

Comment:
I like that you added a test case for this. People often forget to add tests :)

I'd suggest adding another test for when there is only one variable. That is a code path that we are not testing.

    for x in foo()

It would also be nice if we had some test cases for errors, to ensure that it does something sensible there as well. Those can go with those `assert_statements_syntax_error` near the bottom, next to the syntax errors for the ForNum loops. 

    for in ipairs() do -- missing the declarations
    for i x in ipairs do -- missing a comma

By the way, it is better when the error message is similar to Lua but it is different then it is not a big deal.

Reply:
Ah right, I'll add those.